File: src/core/snooze.ts
import { hasNewReplyInMyThreadsSince } from '@core/threads'
import type { PullRequest, Snooze } from '@shared/types'

/**
 * A snooze parks a PR in the "waiting" section. It stays active until its own
 * wake condition fires.
 */
export function isSnoozeActive(
  pr: PullRequest,
  snooze: Snooze,
  myLogin: string,
  now: string,
): boolean {
  switch (snooze.type) {
    case 'until-time':
      return snooze.until !== undefined && now < snooze.until
    case 'until-new-commits':
      return pr.lastCommitPushedAt <= snooze.snoozedAt
    case 'until-reply':
      return !hasNewReplyInMyThreadsSince(pr, myLogin, snooze.snoozedAt)
  }
}

File: Sources/PulloverCore/Snooze.swift
import Foundation

/// A snooze parks a PR in the "waiting" section. It stays active until its own
/// wake condition fires.
public func isSnoozeActive(_ pr: PullRequest, _ snooze: Snooze, myLogin: String, now: Date) -> Bool {
    switch snooze.type {
    case .untilTime:
        guard let until = snooze.until else { return false }
        return now < until
    case .untilActivity:
        return !hasNewReplyInMyThreads(pr, myLogin: myLogin, since: snooze.snoozedAt)
            && !hasNewPush(pr, since: snooze)
    }
}

/// Whether the branch moved since the snooze was set. By SHA when both sides
/// have one: a commit's date is when it was made, and a rebase or cherry-pick
/// pushes commits dated before the snooze. By date only for a snooze saved
/// before the SHA was recorded.
private func hasNewPush(_ pr: PullRequest, since snooze: Snooze) -> Bool {
    if let then = snooze.headSHA, let now = pr.headSHA { return then != now }
    return pr.lastCommitPushedAt > snooze.snoozedAt
}
